### ISSUE TITLE
OpenShot needs to manage source files by reference, not direct path

### ISSUE THREAD

**Issue by ferdnyc** [OpenShot/openshot-qt#1325]
**System Details:**
* Operating System / Distro: Any
* OpenShot Version: 2.4.1

**Issue Description and steps to reproduce:**

I realize this would require a _significant_ overhaul of the project-file format and management within the code, but I think it's a necessary one. Currently, in the project file, OpenShot stores each clip's source as its location on disk, using the direct path. This means that, despite having the concept of "project files" in the interface, each clip is actually completely separate from each other, even if they come from the same source file.

When opening a project with missing files, this causes significant problems for the user because they will be required to re-target not even just each individual missing file, but each **instance** of the missing file. If a clip is used 20 times in a project (not out of the question, for things sound effects), then opening that project with the file missing means that the user will be presented with the "Missing File" dialog and file-chooser **TWENTY TIMES**. Lather, rinse, repeat for each _instance_ of each missing file.

This leads to reports like bug #1320 where OpenShot is perceived to have gone into an infinite loop asking for the location of missing files. It would've eventually completed the loop, but only after the user had either given the location or canceled the dialog over, and over, and over again for each clip instance that uses a file that was not available. That's not acceptable.

If OpenShot embedded the "Project Files" list in the export file, and all clips' locations were stored as references to this list, then worst-case the user would only have to choose the location of each missing file _once_. (Ideally, once they chose a directory path for missing files, all of the references to files in the same directory would be relocated automatically.)

(It's also nearly **impossible** to abort this process, which is a separate issue. Short-term, the Missing File dialog needs both "Skip" and "Cancel" buttons, to prevent repeated file-chooser dialogs from locking the user out of even escaping the load process. On Linux, an `openshot-qt` instance I'd started from the terminal wouldn't even accept Ctrl-C to kill it, while this was going on. I had to pause it with Ctrl-Z and manually `kill` the process to get rid of it.)

**Comment by ferdnyc** [comment]
I actually have a project file that's composed of nothing but _one_ video file which I've pasted into the timeline 15 times as a sort of poor-man's looping, since OpenShot doesn't really have the concept of a "looped file" internally. 

(That would be a great **#someday** enhancement, if a clip could be set to loop indefinitely on the timeline without creating multiple instances, and dragging the clip edges simply adjusted the number of times it looped. But for now that's beside the point.)

If I rename that **one** video file and try to open the project, sure enough I'm presented with the Missing File dialog and file chooser _fifteen_ times, and have to select the same new location _fifteen_ times. For a single missing file!

**Comment by peanutbutterandcrackers** [comment]
This is great suggestion. I think this sure would make OpenShot better. I hope this can be done ASAP.

**Comment by ferdnyc** [comment]
The _only_ time it **might** make sense to use relative paths for project media is when those paths are **INSIDE** the same folder structure that holds the project file itself. That allows the project file and media to be moved, as long as they're moved together.

The common pattern for this sort of thing would be based around storing OpenShot project file(s) in a Project Folder, rather than as a single "project file". There are two approaches to that, though the differences between the two are insignificant: 
* The Audacity-style approach. A single `PROJECTNAME.aup` file is created alongside a `PROJECTNAME_data` folder, which holds the data managed by that project file. Paths in the `.aup` file are implicitly relative to the data folder, except for any (absolute) file data imported by reference, if that ability is enabled in the preferences. (If it's not, then all data is automatically copied into the project folder on import, and there are no external references at all.) 
  The software includes a "Check dependencies..." feature which can report any externally-referenced media, and a "Save compressed copy of project..." target in the File→Export submenu which will create a copy of the project file / data folder pair, with all media not only copied into the data folder but also compressed for most efficient storage.

* The git-style approach (used by almost all coding IDEs and version control systems, really). The "standard unit" for a project is _not_ any particular file, but rather a folder in which the project's data is managed. Control files related to the project will be stored in that folder, either directly at the top level or within hidden subdirectories (`.git/`), depending whether they're intended to be directly examined by the user. All paths are relative to the project folder root, and "copying/moving the project" would be a matter of duplicating or moving that entire _folder_ as a unit. None of the files within are meaningful on their own.

OpenShot certainly _could_ move to a project-folder-based system (though it certainly doesn't _have_ to, and there's no reason the project-file-based approach isn't workable); as I noted in the comments on another issue, using project folders could actually help solve a lot of issues. Probably even ones we haven't even considered yet. (Like, for instance, it would open up the possibility of retaining cache data per-project, instead of caching centrally in `$HOME/.openshot-qt/` and flushing out the entire cache each time a project file is loaded. That would probably require additional logic to manage disk space, though, since right now the short-lived nature of those caches just automatically prevents them from growing very large.) But chief among those advantages would be the ability to (optionally) store imported media inside the folder, so that it's "owned" by the OpenShot project, reducing the likelihood that it's unable to be located during subsequent project sessions. Those files should, of course, be referenced by relative path from the root of the project folder.

**ANYTHING** _outside_ of the project folder should be referenced by absolute path, because moving the project folder should not have the side-effect of _breaking_ those references.

And as long as "the project" is managed as just a single data file, _**nothing**_ should be using relative paths, for a simple reason: It's impossible to have a path relative to a file, only to a folder, and the assumption that the project file will remain located in the same folder is a _terrible_ one. (It makes it impossible to move the project file to another location on disk without breaking **EVERY** reference, as is currently the case with OpenShot `.osp` files.)

**Comment by peanutbutterandcrackers** [comment]
We should probably get this thing done within the 2.4.x series so that we'll be able to share the project files with which the users are encountering the problems... :)

**Comment by dalefrakes** [comment]
Having relative paths for a project would be incredibly helpful.  I have 2 use-cases related to building videos for my online classes.

First, as I build a lecture video, I already put all the files within a project folder, e.g. `week-01-videos`.  After recording bits, screen captures, etc. on my laptop (which is quiet, but slow for renders), I move the project a much more powerful machine for rendering.  It would be nice if the directory structures didn't have to be identical between these machines.

Second, as I go from one term to another, my class folders (including videos) get moved into an "Archive" directory, and a new folder with a date-coded name is made for the next session of the class (e.g. `MSRP-SP2018` to `MSRP-SP2019`.  I'll copy the video projects from the previous year into the new year for editing, but now the path is different and it causes all kinds of problems.  I've tried doing a search & replace in the project definition file, but with limited success.  This gets even more challenging when I'm working in Linux for part of the work, and in Windows for the rest, since the root of the path can't be the same.

For me, "being able to launch openshot and being working without a project defined" is not important at all.  But it seems like it wouldn't be too difficult to be able to define a project path, and have all files stored under that path be "relative".  They could be kept in-memory as absolute paths (so nothing else has be changed), then when they are saved and loaded, the paths within the project folder could be stored "relative".  Maybe make this a  `[ ] store project paths as relative` check box, withe a default to "off" so the default behavior doesn't change.

**Comment by ferdnyc** [comment]
> For me, "being able to launch openshot and being working without a project defined" is not important at all. But it seems like it wouldn't be too difficult to be able to define a project path, and have all files stored under that path be "relative". 

A lot of applications, media editors in particular, don't allow the "dive right in" style of interaction, this being just one of several reasons. If you're not launching from an existing project file, then the first thing you get is a project loading/creation wizard where you do things like set the media parameters for the project (no more working in the wrong export profile because you simply forgot to change it at the beginning) and choose file output locations (and possibly input library path(s) as well).

I definitely see the appeal of the "immediately ready for use" mode of OpenShot's current design, but there are some real tradeoffs to that approach. I guess the question is, does it lead to better or worse experiences for users? (The question is definitely **NOT** "which do users prefer", because in general most users will always say they _prefer_ not being forced to make decisions and deal with dialog boxes. But that doesn't mean they're right.)

**Comment by KERNED** [comment]
Any change this enhancement will be realized soon? Ran into this issue trying to collaborate with someone on Windows (i am on macos) via dropbox... Cross-platform doesn't really seam to matter though as i expect it won't work between two windows machines due to the absolute paths... Right?

**Comment by steve-pence** [comment]
Hi Guys. I know documentation is not cool, especially any documentation that is not "concept free." I stumbled onto this thread trying, during my first fifteen minutes as a OpenShot user, to figure out how in the stink OpenShot is managing project files! Is it copying them somewhere? Does it import them into a database, or a folder in the file system? If not, does it use absolute disk location, relative disk location? This actually really important for users to understand. Conceptually. And I find nothing documented about this. Sorry, if I don't know where to look. I don't even know if this request for by reference file  has been implemented. It sounds like it has not, and that what happens CURRENTLY when I drag a file into the project window is that OpenShot saves its local file path. And whenever that file is referenced, ie dragged onto a time line, a copy is made of that file path. (Ugh!!!LY!!) So it is. 

The only thing worse that it being that way is not knowing that it is that way. Would someone kindly confirm for me OpenShot's current architecture in this area?

Sorry for hijacking this thread, sort of.
Steve

**Comment by N3WWN** [comment]
Looking at my most recent project files from 2.5.1-dev2, it appears that OpenShot is currently using relative paths.  What is odd is that they appear to be based on the `export_path`, not the `import_path`.  

Since these two are the only absolute paths that I see in my project files, I'm going to assume that the `export_path` is the same as the path to the .osp file (essentially the project path) since exporting defaults to the directory that the .osp file is located.

There's a mix of relative path convention, too.

Some clips, the ones in the same dir as the .osp file, have a leading `./` while clips in a subdirectory do not.

Here's a snippet from grepping for path which shows the two conventions:
```
    "path": "./HOA.svg",
    "path": "Shot Coach-Open/IMG_0827_cas85.MP4",
```

I would expect relative paths to be consistently referenced like this:
```
    "path": "./HOA.svg",
    "path": "./Shot Coach-Open/IMG_0827_cas85.MP4",
```

**Comment by jandermoreira** [comment]
This feature would be a major improvement, IMHO.

**Comment by N3WWN** [comment]
I haven't verified in the absolute newest daily builds, but as of OpenShot-v2.6.1-dev-daily-8409-111026d8-73c519f0-x86_64.AppImage, this is still an issue.

**Comment by N3WWN** [comment]
Confirmed issue still exists with recent daily build (OpenShot-v2.6.1-dev-daily-9604-5776efd7-46255e46-x86_64.AppImage).  

It actually looks like the leading `./` is removed for subdirectories under `.`.  I don't have any projects that reference files outside of the project directory (`.`), so can't confirm if the issue extends above `.`.
